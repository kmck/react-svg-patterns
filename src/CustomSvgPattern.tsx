import * as React from 'react';

interface Props {
  children?: JSX.Element | JSX.Element[] | ((id: string) => JSX.Element) | null,
  id: string,
};

const CustomSvgPattern = ({
  children = null,
  id,
}: Props) => {
  return (
    <pattern
      height={1}
      id={id}
      patternContentUnits="objectBoundingBox"
      patternUnits="objectBoundingBox"
      width={1}
    >
      {typeof children === 'function' ? children(id) : children}
    </pattern>
  );
};

export default React.memo(CustomSvgPattern);
