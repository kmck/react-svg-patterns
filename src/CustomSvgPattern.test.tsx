import * as React from 'react';
import renderer from 'react-test-renderer';

import CustomSvgPattern from './CustomSvgPattern';

describe('CustomSvgPattern', () => {
  it('renders a custom SVG pattern', () => {
    const tree = renderer.create(
      <CustomSvgPattern id="test" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a custom SVG pattern with a single child', () => {
    const tree = renderer.create(
      <CustomSvgPattern id="test">
        <rect x={0} y={0} width={1} height={1} fill="#00f" />
      </CustomSvgPattern>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a custom SVG pattern with multiple children', () => {
    const tree = renderer.create(
      <CustomSvgPattern id="test">
        <rect x={0} y={0} width={1} height={1} fill="#ff0" />
        <rect x={0} y={0} width={0.5} height={0.5} fill="#f00" />
        <rect x={0.5} y={0.5} width={0.5} height={0.5} fill="#f00" />
      </CustomSvgPattern>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a custom SVG pattern with a render prop', () => {
    const tree = renderer.create(
      <CustomSvgPattern id="test">
        {id => (
          <g>
            <rect x={0} y={0} width={1} height={1} fill="#ff0" />
            <rect x={0} y={0} width={0.5} height={0.5} fill="#f00" />
            <rect x={0.5} y={0.5} width={0.5} height={0.5} fill="#f00" />
            <text>{id}</text>
          </g>
        )}
      </CustomSvgPattern>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
