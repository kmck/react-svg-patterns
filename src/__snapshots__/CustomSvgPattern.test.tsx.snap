// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomSvgPattern renders a custom SVG pattern 1`] = `
<pattern
  height={1}
  id="test"
  patternContentUnits="objectBoundingBox"
  patternUnits="objectBoundingBox"
  width={1}
/>
`;

exports[`CustomSvgPattern renders a custom SVG pattern with a render prop 1`] = `
<pattern
  height={1}
  id="test"
  patternContentUnits="objectBoundingBox"
  patternUnits="objectBoundingBox"
  width={1}
>
  <g>
    <rect
      fill="#ff0"
      height={1}
      width={1}
      x={0}
      y={0}
    />
    <rect
      fill="#f00"
      height={0.5}
      width={0.5}
      x={0}
      y={0}
    />
    <rect
      fill="#f00"
      height={0.5}
      width={0.5}
      x={0.5}
      y={0.5}
    />
    <text>
      test
    </text>
  </g>
</pattern>
`;

exports[`CustomSvgPattern renders a custom SVG pattern with a single child 1`] = `
<pattern
  height={1}
  id="test"
  patternContentUnits="objectBoundingBox"
  patternUnits="objectBoundingBox"
  width={1}
>
  <rect
    fill="#00f"
    height={1}
    width={1}
    x={0}
    y={0}
  />
</pattern>
`;

exports[`CustomSvgPattern renders a custom SVG pattern with multiple children 1`] = `
<pattern
  height={1}
  id="test"
  patternContentUnits="objectBoundingBox"
  patternUnits="objectBoundingBox"
  width={1}
>
  <rect
    fill="#ff0"
    height={1}
    width={1}
    x={0}
    y={0}
  />
  <rect
    fill="#f00"
    height={0.5}
    width={0.5}
    x={0}
    y={0}
  />
  <rect
    fill="#f00"
    height={0.5}
    width={0.5}
    x={0.5}
    y={0.5}
  />
</pattern>
`;
